### ISSUE TITLE
Proposal: RPC protocol for tari_comms

### ISSUE THREAD

**Issue by sdbondi** [tari-project/tari#2092]
This proposal is (tick applicable):

* [x] A new feature
* [ ] An extension to an existing RFC/proposal <LINK>
* [ ] An alternative approach to an existing RFC or proposal <LINK>

### Change log

| Date | Changes |
|:-----|:--------|
| 28/07/2020     | Initial        |

### Goals
 
_Fire-and-forget_ messaging is used throughout the current system. While some areas of the system suit this approach perfectly (transaction/block/DHT propagation), it falls short for many current use-cases.

This proposal aims to address the shortcomings of the current messaging approach, namely:
- excessive complexity (which causes bugs and code that is hard to maintain)
- lack of separation of concerns (nodes have to worry about messaging semantics)
- performance is difficult (the current protocol lends itself to implementers constructing large messages and makes _streaming_ many small messages difficult)
- visibility is difficult (not knowing if a message has actually been received. Knowing if a peer is actually connected requires holding a handle and querying connectivity manager)
- Writing protocols (conversations) is difficult. Any type of message can arrive at any time. This means that the domain layer has to implement a mapping between requests and responses. Additionally, defending against unexpected, potentially malicious messages is something that has to be actively reasoned about for every incoming message type (Did I request this message? Did this message come from the peer I expected it to?). 
- Unbounded tasks are often spawned per message, which can be an attack vector
- Testing is difficult. Outbound and inbound messaging is difficult to stub out.

### Assumptions

- A reliable transport such as TCP is used
- Many devs are familiar with gRPC. 
- Implementing a RPC interface similar to tonic (except p2p) using the current comms infrastructure will not be prohibitively difficult.

### Abstract

The proposes a more bounded and structured approach to protocols on the Tari p2p network heavily influenced by tonic RPC.

What the proposed RPC protocol does do:
- mapping requests to responses
- managing timeouts and disconnects
- message framing
- Rate limiting
- consistent `RpcError` api
- (de)serialization of requests/responses (protobuf)
- stream requests and/or responses
- async service and client trait (async_trait crate)

What an RPC protocol does not do / limitations:
- Peer discovery, store and forward, message propagation/broadcasting (use the DHT/outbound messaging handles)
- Reconnecting disconnected peers (Use connectivity manager)
- Resending messages or any "retry behaviour" (the request either works or doesnt)
- Requests without responses

Future extensions/nice to haves 
- use protobuf service definitions, that use rust proc macros and prost for code generation to reduce boilerplate
- Auto banning of misbehaving peers (if configured to do so)
- Message validators 

### Some background on protocol negotiation

`tari_comms` supports independent p2p channels, called substreams. The peer that initiated the substream MUST send the protocol that should be spoken. If the destination knows this protocol, the substream is handed over to the preregistered handler. This process is abstracted by the `PeerConnection::open_substream` call. Since all peer connections are authenticated (that is, we're proven that the peer has a private key of the public key and the node id is a hash of the public key), there is no need to implement RPC authentication.

In the context of RPC, these substreams would be leveraged so that there is no need to write code to split out the various messages and dispatch them to the correct handler. Moreover, a protocol identifier (e.g. `/tari/mempool/1.0/rpc/1`) would not need to be sent along with each message since it is negotiated once ahead of time. Substreams would be opened lazily as requested by peers.

### RPC architecture

![Untitled Diagram](https://user-images.githubusercontent.com/1057902/88630629-2e222580-d0b1-11ea-9544-a2cdbbf3c09e.png)

### Request

```proto
message RpcRequest {
   // Request ID
    uint32 request_id = 1;
    // Method identifier
    uint32 method = 2;
    // Request flags
    uint32 flags = 3;
    // Message payload
    bytes message = 10;
}
```

Given that the peer and protocol are negotiated ahead of time, this information does not need to be included in requests.
The `request_id` and `method` fields can be restricted to `u16` (max `65,535` / 3 bytes in varint), the overhead per request message is 6-10 bytes (incl 4 bytes for protobuf field markers).

**Flags**

TBD: applies to request and response and basically just mark if there are more messages to come when streaming. All "normal" RPC messages will use `Fin` flag (i.e. empty, 0 bytes for flag field). Another 7 bitflags are reserved for future use.

- `Fin = 0x00` 
- `More = 0x01` (more messages to follow that use the same request ID)

### Response

```proto
message RpcResponse {
   // Id matching the request
    uint32 request_id = 1;
   // The response status (see "Response statuses" below)
    uint32 status = 2;
    // An optional description of an error
    string error_desc = 3;
    // Response flags
    uint32 flags = 4;
    // Message payload
    bytes message = 10;
}
```

**Response Statuses**

```
// Request succeeded
Ok = 0
// Request is incorrect
BadRequest = 1
// The method does not exist
MethodNotFound = 2
// Misc. errors
General = 3
// Method is not implemented
NotImplemented = 4
// The request ID conflicted with another in-flight request
Conflict = 5,
// The timeout was reached before a response was received (client only)
Timeout = 6
```

Low values are selected (as opposed to http codes, 404, 500, 200 etc) to reduce the overhead of the status field when varint encoded (values 1-128 are encoded in 1 byte, 0 in 0 bytes).

**Request ID**

Request IDs are assigned incrementally by the requester and are scoped by the peer substream and protocol - meaning that request IDs do not conflict with other peer/protocols. The receiving protocol MAY return a Conflict error if a request id conflicts with an in-flight request for the same peer/protocol. Request IDs SHOULD wrap around to 0 once 65,535 (`u16::MAX`) is reached. Client request IDs MUST NOT exceed `u32::MAX`.

Any responses received for invalid/timed out requests are logged and discarded.

**Steaming**

Streaming, in the RPC context, means continuously sending messages to a peer without waiting for the peer to request each message. This can be implemented over-the-wire by reusing the request ID and setting the `More` flag. The RpcService will forward messages to a StreamReceiver.

**Public API**

Heavily influenced by tonic. Initially there may be some boilerplate to write that matches request method ids to methods in the service trait (e.g. `MempoolService`). This will be implemented using `tower::Service`.

Service: 

```rust
pub struct MyMempoolService {
    mempool: Mempool,
}

#[tari_comms::async_trait]
impl generated::MempoolService for MyMempoolService {
    const PROTOCOL_NAME: Bytes = Bytes::from_static(b"/tari/mempool/1.0");

    async fn submit_transaction(&self, context: RpcContext<'_>, request: Request<Transaction>) -> Result<Response<()>, RpcStatus> {
        let peer = context.peer_node_id();
        Err(RpcStatus::not_implemented())
    }

    async fn receive_transactions(&self, _ctx: RpcContext<'_>, request: Request<Streaming<Hash>>) -> Result<Response<Streaming<Transaction>>, RpcStatus> {
        let (tx, rx) = mpsc::channel(10);
        task::spawn(async move {
            while let Some(hash) = request.next_message().await {
                let txn = self.mempool.get(hash);
                tx.send(txn).await?;
            }
            Ok(())
        });

        Ok(Response::new(rx.into()))
    }
}
```
Client:

```rust
let peer_conn = connectivity_manager.dial_peer(node_id).await?;
let client  = peer_conn.connect_rpc::<generated::MempoolClient>().await?;
client.submit_transaction(txn).await?;

let (tx, rx) = mpsc::channel(1);
let resp = client.receive_transactions(rx).await?;

// Can spawn a task to send and receive concurrently
task::spawn(async move {

  for hash in hashes {
       tx.send(hash).await;
  }
});

while let Some(txn) = resp.next_message().await {
   // do something with txn
}


```

### New dependencies (`rpc` feature flag)

- [async_trait](https://crates.io/crates/async-trait)
- [tower](https://crates.io/crates/tower) - already a dependency for tari_comms_dht

**Comment by stringhandler** [comment]
I agree with the idea of simplifying the RPC interface. I like having the simplified service and having a context available. I am a little bit worried that there is a lot of work here, and that we don't know for sure if the interface will be simpler in the end. That's not a "no", but let me give this a bit more thought

**Comment by sdbondi** [comment]
Closed by #2121 #2162 #2142
